sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -287,6 +288,17 @@ object ColumnPruning extends Rule[LogicalPlan] {
 }
 
 /**
+ * Combines the adjacent [[Project]] and [[Join]] operators, iff their output sets are the same.
+ */
+object CombineProjectJoin extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case p @ Project(_, j: Join)
+      if j.outputSet == p.outputSet => j
```

Comment:
@cloud-fan I did a try, but it touches a lot of codes. We need to add a new parameter `projectList` into `Join`. I am not sure if this is a right direction. What do you think? 

As what @rxin said, we might can do that for all the operators. That means, we need to add a new parameter `projectList` into all the operators.